apply/shrink.go

```diff
@@ -0,0 +1,117 @@
+// Copyright © 2021 Alibaba Group Holding Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apply
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/alibaba/sealer/common"
+	v1 "github.com/alibaba/sealer/types/api/v1"
+)
+
+type Shrink struct {
+}
+
+func (p *Shrink) Scale(cluster *v1.Cluster, scalingArgs *common.RunArgs) error {
+	switch cluster.Spec.Provider {
+	case common.BAREMETAL:
+		if err := PreProcessIPList(scalingArgs); err != nil {
+			return err
+		}
+		if (!IsIPList(scalingArgs.Nodes) && scalingArgs.Nodes != "") || (!IsIPList(scalingArgs.Masters) && scalingArgs.Masters != "") {
```

Comment:
The logic can be simpler, like this:
```suggestion
if IsIPList(nodes) && nodes != "" {
...
   cluster.spec.nodes = removeIPListDuplicatesAndEmpty(nodes)
}
if IsIPList(masters) && masters != "" {
...
    cluster.spec.master = removeIPListDuplicatesAndEmpty(nodes)
}
```

Reply:
adopt！